File: DailyChallenges/Scala/LeetCode_1061.scala
object Solution {
    def smallestEquivalentString(s1: String, s2: String, baseStr: String): String = {
        val v = (0 until 26).toArray;

        def find(i: Int): Int = if (v(i) == i) i else {
            v(i) = find(v(i)); v(i)
        }

        for (i <- 0 until s1.length){
            val i1 = find(s1(i) - 'a')
            val i2 = find(s2(i) - 'a')
            if (i1 > i2) v(i1) = i2 else v(i2) = i1
        }

        baseStr.map((ch: Char) => ('a' + find(ch - 'a')).toChar)
    }
}

File: DailyChallenges/CPP/LeetCode_1061.cpp
class Solution {
public:
    string smallestEquivalentString(string s1, string s2, string baseStr) {
        vector<int> v(26); 
        iota(v.begin(), v.end(), 0);
        function<int(int)> find = [&](int i) { return (v[i] == i) ? i : (v[i] = find(v[i]));  };
        
        int n = s1.size();
        for (int i=0; i<n; ++i){
            int i1 = find(s1[i] - 'a');
            int i2 = find(s2[i] - 'a');
            if (i1 > i2) swap(i1, i2);
            v[i2] = i1;
        }

        for (int i=0; i< baseStr.size(); ++i){
            int j = find(baseStr[i] - 'a');
            baseStr[i] = 'a' + j;
        }
        
        return baseStr;
    }
};
